pylabrobot/resources/plate_adapter.py

```diff
@@ -0,0 +1,175 @@
+        category (Optional[str], optional): The category of the PlateAdapter.
+          Defaults to "plate_adapter".
+        model (Optional[str], optional): The model of the PlateAdapter. Defaults to None.
+   
+  Examples:
+    1. Using a "magnetic rack" as a PlateAdapter:
+
+      Define a `MFXCarrier`, define a `MFXModule` for deep well plates,
+      assign the DWP module to the MFXCarrier:
+      >>> mfx_carrier_1 = MFX_CAR_L5_base(name='mfx_carrier_1')
+      >>> mfx_carrier_1[1] = mfx_dwp_module_1 = MFX_DWP_module(name="mfx_dwp_module_1")
+
+      Define Alpaqua magnet as a PlateAdapter & assign to mfx_dwp_module_1:
+      >>> Alpaqua_magnum_flx_1 = Alpaqua_magnum_flx(name='Alpaqua_magnum_flx_1')
+      >>> mfx_dwp_module_1.assign_child_resource(Alpaqua_magnum_flx_1)
+
+      Define Cos_96_Rd plate & assign to PlateAdapter:
+      >>> Cos96_plate_1 = Cos_96_Rd(name='Cos96_plate_1')
+      >>> Alpaqua_magnum_flx_1.assign_child_resource(Cos96_plate_1)
+  """
+
+  def __init__(
+    self,
+    name: str,
+    size_x: float, size_y: float, size_z: float,
+    dx: float, dy: float, dz: float,
+    adapter_hole_size_x: float,
+    adapter_hole_size_y: float,
+    site_pedestal_z: float,
+    adapter_hole_dx: float = 9.0,
+    adapter_hole_dy: float = 9.0,
+    category: Optional[str] = None,
+    model: Optional[str] = None
+    ):
+    if adapter_hole_dx not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dx must be one of 9.0, 4.5, or 2.25")
+    if adapter_hole_dy not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dy must be one of 9.0, 4.5, or 2.25")
+
+    super().__init__(name=name, size_x=size_x, size_y=size_y, size_z=size_z,
+      category=category or "plate_adapter", model=model)
+
+    self._child_resource_location = None
+    self._child_resource: Optional[Resource] = None
+    self.dx = dx
+    self.dy = dy
+    self.dz = dz
+    self.adapter_hole_size_x = adapter_hole_size_x
+    self.adapter_hole_size_y = adapter_hole_size_y
+    self.adapter_hole_size_z = self.get_size_z() - self.dz
+    self.adapter_hole_dx = adapter_hole_dx
+    self.adapter_hole_dy = adapter_hole_dy
+    self.site_pedestal_z = site_pedestal_z
+
+  @property
+  def child_resource_location(self) -> Optional[Coordinate]:
+    return self._child_resource_location
+
+  def assign_child_resource(
+    self,
+    resource: Plate,
+    location: Optional[Coordinate] = None,
+    reassign: bool = True
+  ):
+    """Assign a Plate to a PlateAdapter. If `location` is not provided,
+    the resource will autoadjust the placement location based on the
+    PlateAdapter-Plate relationship."""
+
+    if self._child_resource is not None and not reassign:
+      raise ValueError(f"{self.name} already has a child resource assigned")
+    if not isinstance(resource, Plate):
+      raise ValueError("Only plates can be assigned to Alpaqua 96 magnum flx.")
+
+    # TODO: have discussion oon whether to transfer flat bottom error checking
```

Comment:
Can we simply check `well.bottom_type != WellBottomType.flat`?

Reply:
I am not sure: There might be flat-bottom wells that people want to place on a flat-bottom plate_adapter (e.g. for temperature control and/or shaking)?

Reply:
Ah. I thought PlateAdapter here always has holes. But perhaps this is intended to be a more general class? What are the attributes like .adapter_hole_size_x in such cases?

Reply:
Yes, I think to house a `Plate` there will always be holes. But the holes can follow the geometry of any well.bottom that exists.
Bioshake has a [really nice website here](https://www.bioshake.com/collections/adapter), showcasing a variety of adapters that are designed for efficient temperature transfer and/or efficient shaking transmission into the plate.

I am not sure because I have not seen a flat-bottom well plate inside a flat-bottom adapter but I am not sure I want to exclude the option here.
Tbh placing a U-bottom plate inside a V-bottom adapter is in that sense also not a great idea... I'm not sure this kind of safety check can easily be implemented though.

Reply:
how about [this](https://www.bioshake.com/collections/adapter/products/1808-1021-adapter-for-microplate-flat-bottom-standard) from the linked website? I'm not sure if this is the level of complexity we need to take in from the start, but perhaps worth discussing how we might think about this in the future.